### ISSUE TITLE
losses went into inf/nan if optimizer is not SGD

### ISSUE THREAD

**Issue by Peldom** [sokrypton/ColabDesign#6]
I tested other optimizers like adam/rmsprop/momentum from '[jax.example_libraries.optimizers'](https://github.com/google/jax/blob/36cd56824def5a6caac2cf1b837f676cfbc8dfb7/jax/example_libraries/optimizers.py). As design initiated, the whole structure predicted by AlphaFold2 became a “black hole” after 10-20 steps, then loss value changed into inf/nan after 60-80 steps, yet the inputs seems normal. I tried several times per optimizer and got the same result.

**Comment by sokrypton** [comment]
Oops. I fixed this issue in the beta branch, but not yet in the main branch.

Can you try again?

**Comment by Peldom** [comment]
Thank you sir, that was fast. Still the "**black hole**" bug exists(especially in the soft_iters, there is a sharp decline and then a sharp increase in losses), seems Evofomer learnt chaos from  **self._grad** & **self._state** that cannot provide meaningful information to structure module.

**Comment by sokrypton** [comment]
Hi Peldom, can you share the example you are working with? I'll investigate.

**Comment by Peldom** [comment]
key codes: (adam, rmsprop, etc)
```py
protocol == "binder"
from jax.example_libraries.optimizers import rmsprop
model._setup_optimizer(optimizer=rmsprop)
model.design_3stage(soft_iters=40, temp_iters=40, hard_iters=20)
```
And the structure is predicted like this when all the losses are quiet low(around step 20)
![black_hole_bug](https://pic4.zhimg.com/80/v2-a7d3ef5c2b674cf777ce1bbeb2072967_720w.jpg)
Take 'pae_intra' for example:
![WechatIMG518](https://user-images.githubusercontent.com/51025363/163423849-ab34ee61-c651-44fa-829a-bf5e688c263b.png)

I'm so confused😅 , it happens more often in **soft_iters** than **temp_iters/hard_iters**,  I would be so grateful if you could find out why

**Comment by sokrypton** [comment]
Did you change how the sequence was initialized? Or is this the result of changing the optimizer?

The current optimizer was calibrated (zero initialization, dropouts, switching between model params, normalized gradients) to avoid these adversarial modes.

Can you share the example you are trying on? (if it's private, you can email me at so@fas.harvard.edu).
